### PR TITLE
Show all per class survey results for admins and add teacher search form

### PR DESCRIPTION
**Pull request by willgearty**
Before we only showed numerical per class survey results. Now we include the non-numerical results as well. Also, the results are now grouped into individual and per class sections. I also tweaked some of the styling of the page overall, added a table of contents to the top, and added a search box to select a teacher by which to filter the survey results.

Fixes https://github.com/learning-unlimited/ESP-Website/issues/2938.

### CODE REVIEW

File: esp/templates/survey/review.html [hunk 1]
```diff
@@ -12,19 +12,68 @@
...
+    </li>
+{% endfor %}
+
+{% if perclass_data %}
+{% for c in perclass_data %}
+    {% ifchanged c.class.parent_class %}
+        {% if not forloop.first %}
+            </ul>
+        {% endif %}
+        <li><a href="#class{{ c.class.parent_class.id }}">Responses for {{ c.class.parent_class.emailcode }}: {{ c.class.title }}</a></li>
```
Comment (kkbrum) [inline esp/templates/survey/review.html:55]: Hmm for me all the perclass data for a question shows up in one list of responses, not separated by class, even though the code looks like there are class headings? I get the class headings when I'm filtered to a teacher, but not on the overall page.

File: esp/templates/survey/review.html [hunk 2]
```diff
@@ -12,19 +12,68 @@
...
 </style>
-<br /><br />
-<h1>Survey Responses for {{ program.niceName }} </h1>
+<a name="top"></a><h1>Survey Responses for {{ program.niceName }} </h1>
+
+{% if teacher_form %}
+<form method="POST" action="{{ request.get_full_path }}">
+<h2>Search for a teacher to retrieve their specific survey results:</h2>
+{{ teacher_form.target_user }}
+<input type="submit" value="Search" />
```
Comment (kkbrum) [inline esp/templates/survey/review.html:21]: If I hit "Search" with no teacher entered (yes this is not the point but still), I get the following error:

image

File: esp/templates/survey/review.html [hunk 3]
```diff
@@ -34,9 +83,10 @@ <h1>Survey Responses for {{ program.niceName }} </h1>
 {% if perclass_data %}
 {% for c in perclass_data %}
```
Comment (kkbrum) [inline esp/templates/survey/review.html:84]: Could we have each separate class on the results page filtered to a teacher have some sort of nice sections for the classes like either a border between them or boxes or something? It looks kinda plain with just the headings.

File: esp/templates/survey/review.html [hunk 4]
```diff
@@ -34,9 +83,10 @@ <h1>Survey Responses for {{ program.niceName }} </h1>
...
     <table class="fullwidth">
-    <tr>
-        <th class="small">Responses for {{ c.class.emailcode }}: {{ c.class.title }}</th>
-    </tr>
+        {% ifchanged c.class.parent_class %}
+        <tr><th><h2><a name="class{{ c.class.parent_class.id }}"></a>Responses for {{ c.class.parent_class.emailcode }}: {{ c.class.title }}<span style="float:right;font-weight: normal;"><a href="#top">[to top]</a></span></h2></th></tr>
```
Comment (kkbrum) [inline esp/templates/survey/review.html:87]: Could the class name heading be larger than the section name heading? It feels weird to me that they are equal font sizes.

File: esp/templates/survey/review.html [hunk 5]
```diff
@@ -12,19 +12,68 @@
 <style type="text/css">
 .nocheckmark { border: 1px solid black; }
 </style>
-<br /><br />
-<h1>Survey Responses for {{ program.niceName }} </h1>
+<a name="top"></a><h1>Survey Responses for {{ program.niceName }} </h1>
+
+{% if teacher_form %}
```
Comment (kkbrum) [inline esp/templates/survey/review.html:17]: I'm morphed into a non-admin teacher and I still see the search form.

Comment (kkbrum) [reply esp/templates/survey/review.html:17]: (also tried just logging into this non-admin teacher and same thing so don't blame the morphing :P)

File: esp/templates/survey/review.html [hunk 6]
```diff
@@ -12,19 +12,68 @@
...
+<form method="POST" action="{{ request.get_full_path }}">
+<h2>Search for a teacher to retrieve their specific survey results:</h2>
```
Comment (kkbrum) [inline esp/templates/survey/review.html:19]: Can there be a way to clear the search for the teacher and return to the non-filtered results page?

File: esp/templates/survey/review.html (hunk 5 above)
Comment (willgearty) [reply esp/templates/survey/review.html:17]: Yeah, I noticed this when I was testing. I'm honestly not sure what's going on. I'll look into it again.

File: esp/templates/survey/review.html (hunk 1 above)
Comment (willgearty) [reply esp/templates/survey/review.html:55]: Oh, I hadn't thought about that. Umm... I'll see if it's possible to separate by class for this section.

File: esp/templates/survey/review.html (hunk 5 above)
Comment (willgearty) [reply esp/templates/survey/review.html:17]: I believe this should now be fixed. Let me know if you're still having issues.

**Comment by willgearty** [comment]
OK, all that's left to do is to restyle the page with borders or something (as requested by @kkbrum) and then copy all template changes to the LaTeX template.

File: esp/templates/survey/review.html [hunk 7]
```diff
@@ -11,41 +11,131 @@
...
+        {% if not forloop.first %}
+        </table></br></br>
+        {% endif %}
+        <table class="fullwidth bordered">
+        <tr><th><h2><a name="class{{ c.class.parent_class.id }}"></a>Responses for {{ c.class.parent_class.emailcode }}: {{ c.class.title }}<span style="float:right;font-weight: normal;"><a href="#top">[to top]</a></span></h2></th></tr>
+        {% endifchanged %}
+        <tr style="border-top: 1px dashed black;"><th><h3><a name="class{{ c.class.parent_class.id }}-{{ c.class.index }}"></a>Responses for Section {{ c.class.index }}<span style="float:right;font-weight: normal;"><a href="#top">[to top]</a></span></h3></th></tr>
+        {% for q in c.questions %}
+            {% with c.class.num_students as num_participants %}
+                <tr><td>{% include "survey/answers.html" %}</td></tr>
```
Comment (kkbrum) [inline esp/templates/survey/review.html:110]: For the per class questions in the html template, I'm only getting the "statistics:  Number of responses __/__" line for the first out of the per class questions for each class. In the pdf version, i get this statistic for each individual question.

File: esp/templates/survey/review.tex [hunk 8]
```diff
@@ -11,25 +11,37 @@
...
     {% endfor %}
     \end{longtable}
     \end{center}
+    {% for c in s.perclass_data %}
+        \clearpage
+        \begin{center}
+        \begin{longtable}{|p{0.5\linewidth}|p{0.5\linewidth}|} \hline
+        \multicolumn{2}{|C{\linewidth}|}{\cellcolor{esphead} ~ } \\
+        \multicolumn{2}{|C{\linewidth}|}{\cellcolor{esphead} \Large {Responses for {{ c.class.emailcode }}: {{ c.class.title|texescape }}} } \\ 
+        \multicolumn{2}{|C{\linewidth}|}{\cellcolor{esphead} ~ } \\\hline
```
Comment (kkbrum) [inline esp/templates/survey/review.tex:35]: Can we get the blue box to cover the entire width of the table (right now its a little short, width wise):
image

File: esp/templates/survey/review.tex [hunk 9]
```diff
@@ -11,25 +11,37 @@
...
+        \multicolumn{2}{|C{\linewidth}|}{Teachers: {{ c.class.parent_class.getTeacherNames|join:", "|texescape }} } \\ \hline
+        {% for q in c.questions %}
```
Comment (kkbrum) [inline esp/templates/survey/review.tex:37]: Can you add a little padding above the question name? and above the teacher names while you're at it too... just to be nicer on the eyes. 
image   image

Comment (kkbrum) [reply esp/templates/survey/review.tex:40]: I added some padding-- lmk if it looks good on your end. Note that the histogram is still not well padded but when i tried to add padding it threw off the other column and such so I just left it for now. The other stuff should look much better though.

File: esp/templates/survey/review.tex (hunk 8 above)
Comment (kkbrum) [reply esp/templates/survey/review.tex:35]: Done!

File: esp/templates/survey/review.html (hunk 7 above)
Comment (willgearty) [reply esp/templates/survey/review.html:110]: As the code currently stands, only countable questions will have the "Statistics" section included. Are your other questions short/long answer?

Comment (willgearty) [reply esp/templates/survey/review.html:110]: OK, we now show the number of responses for every question, but I've moved it to next to the title (and now the other statistics are only shown if the question is numeric).

**Pull request merged**